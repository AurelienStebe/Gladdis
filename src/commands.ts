import path from 'path'
import fs from 'fs-extra'
import { JSDOM } from 'jsdom'
import fetch from 'node-fetch'
import TurndownService from 'turndown'
import { getDocument } from 'pdfjs-dist'
import { gfm } from 'turndown-plugin-gfm'
import { arrayBuffer } from 'stream/consumers'

import { doGladdis } from './gladdis.js'
import { transcribe } from './utils/whisper.js'
import { parseLinks } from './utils/scanner.js'
import { webBrowser } from './utils/browser.js'
import { getTokenModal } from './utils/loggers.js'
import { loadContext, loadContent } from './utils/loaders.js'

import type { ReadStream } from 'fs'
import type { PDFDocumentProxy } from 'pdfjs-dist'
import type { Context, DiskInterface } from './types/context.js'

export { stringify as stringifyYaml, parse as parseYaml } from 'yaml'

export function parseDom(html: string): Document {
    return new JSDOM(html).window.document
}

export async function getHtml(url: string): Promise<string> {
    return await (await fetch(url)).text()
}

export function turndown(html: string): string {
    const turndown = new TurndownService({
        hr: '---',
        headingStyle: 'atx',
        bulletListMarker: '-',
        codeBlockStyle: 'fenced',
    })

    turndown.use(gfm)

    return turndown.turndown(html)
}

export async function getPdfDoc(stream: ReadStream | Promise<File>): Promise<PDFDocumentProxy> {
    const file = stream instanceof Promise ? (await stream).arrayBuffer() : arrayBuffer(stream)

    return await getDocument(await file).promise
}

export const diskInterface: DiskInterface = {
    readFile: async (p: string) => await fs.readFile(p, 'utf-8'),
    readBinary: async (p: string) => await arrayBuffer(fs.createReadStream(p)),
    appendFile: async (p: string, d: string) => await fs.appendFile(p, d),
    deleteFile: async (p: string) => await fs.remove(p),
    pathExists: async (p: string) => await fs.pathExists(p),
    pathEnsure: async (p: string) => await fs.ensureDir(p),
    baseName: (p: string, e?: string) => path.basename(p, e),
    extName: (p: string) => path.extname(p),
    joinPath: (...p: string[]) => path.join(...p),
    dirName: (p: string) => path.dirname(p),
}

export async function chatWithGladdis(context: Context): Promise<void> {
    await doGladdis(loadContent(await loadContext(prepareContext(context))))
}

export async function processContent(context: Context): Promise<void> {
    context = loadContent(await loadContext(prepareContext(context)))

    context.whisper.echoOutput = true
    context.whisper.deleteFile = false

    await Promise.all(
        context.user.history.map(async (message) => {
            message.content = await parseLinks(message.content, context)
        }),
    )

    context.user.prompt = await transcribe(context.user.prompt, context)
    context.user.prompt = await parseLinks(context.user.prompt, context)
    context.user.prompt = await webBrowser(context.user.prompt, context)

    context.user.history.push({
        role: 'user',
        content: context.user.prompt,
        name: context.user.label,
    })

    await context.file.disk.appendFile(context.file.path, getTokenModal(context))
}

export async function processPrompt(context: Context): Promise<void> {
    context = await loadContext(prepareContext(context))
    context.file.text = context.file.text.split('\n---').at(-1) ?? ''
    context = loadContent(context)

    context.whisper.echoOutput = true
    context.whisper.deleteFile = false

    context.user.prompt = await transcribe(context.user.prompt, context)
    context.user.prompt = await parseLinks(context.user.prompt, context)
    context.user.prompt = await webBrowser(context.user.prompt, context)

    context.user.history = [
        {
            role: 'user',
            content: context.user.prompt,
            name: context.user.label,
        },
    ]

    await context.file.disk.appendFile(context.file.path, getTokenModal(context))
}

export function prepareContext(context: Context): Context {
    context.user = context.user ?? {}

    context.user.env = context.user.env ?? process.env
    context.file.disk = context.file.disk ?? diskInterface

    return context
}
