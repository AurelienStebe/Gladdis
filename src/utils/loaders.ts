import { deepmerge } from 'deepmerge-ts'

import { parseYaml } from '../commands.js'
import { parseHistory } from './history.js'
import { writeMissedModal } from './loggers.js'

import type { Context } from '../types/context.js'

const linkRegex = /(?<!<%.*)(?<!!)(\[\[[^\]]+?\]\])(?!.*%>)/gs

export async function loadContext(context: Context): Promise<Context> {
    const callContext = deepmerge(context, { file: { date: new Date() } })
    const fileContext = deepmerge(await loadMarkdown(context), callContext)

    const coreContext = {
        file: {
            name: context.file.disk.baseName(context.file.path, '.md'),
            text: '',
        },
        user: {
            data: context.user.env.GLADDIS_DATA_PATH ?? '__DATA__',
            label: context.user.env.GLADDIS_DEFAULT_USER ?? 'User',
            prompt: '',
            history: [],
        },
        gladdis: {
            label: context.user.env.GLADDIS_NAME_LABEL ?? 'Gladdis',
            config: context.user.env.GLADDIS_CONFIG_FILE,
            model: context.user.env.GLADDIS_DEFAULT_MODEL ?? 'gpt-3.5-turbo',
            temperature: Number(context.user.env.GLADDIS_TEMPERATURE ?? 0),
            top_p_param: Number(context.user.env.GLADDIS_TOP_P_PARAM ?? 100),
            freq_penalty: Number(context.user.env.GLADDIS_FREQ_PENALTY ?? 0),
            pres_penalty: Number(context.user.env.GLADDIS_PRES_PENALTY ?? 0),
        },
        whisper: {
            input: context.user.env.GLADDIS_WHISPER_INPUT ?? 'Gladdis',
            config: context.user.env.GLADDIS_WHISPER_CONFIG,
            model: context.user.env.GLADDIS_WHISPER_MODEL ?? 'whisper-1',
            liveSuffix: context.user.env.GLADDIS_WHISPER_LIVE_SUFFIX ?? 'dictated, but not read',
            readSuffix: context.user.env.GLADDIS_WHISPER_READ_SUFFIX ?? 'transcribed and read',
            temperature: Number(context.user.env.GLADDIS_WHISPER_TEMPERATURE ?? 0),
            echoOutput: (context.user.env.GLADDIS_WHISPER_ECHO_OUTPUT ?? 'true').toLowerCase() === 'true',
            deleteFile: (context.user.env.GLADDIS_WHISPER_DELETE_FILE ?? 'false').toLowerCase() === 'true',
            language: context.user.env.GLADDIS_WHISPER_LANGUAGE_ID,
        },
    }

    const fullContext = deepmerge(coreContext, fileContext) as unknown as Context
    return deepmerge(await loadAIConfig(fullContext), fileContext) as unknown as Context
}

export async function loadAIConfig(context: Context): Promise<Context> {
    const disk = context.file.disk
    let configContext: any = { whisper: {} }

    if (context.gladdis.config !== undefined && context.gladdis.config !== '') {
        if (!context.gladdis.config.toLowerCase().endsWith('.md')) context.gladdis.config += '.md'
        const configPath = disk.joinPath(context.user.data, 'configs', context.gladdis.config)

        if (await disk.pathExists(configPath)) {
            configContext = deepmerge(configContext, {
                file: { path: configPath, disk },
                user: { label: 'System' },
            })

            configContext = await loadMarkdown(configContext as Context)
            const configHistory = deepmerge(context, configContext) as Context

            delete configContext.file
            delete configContext.user

            context = deepmerge(context, configContext)
            context.user.history.unshift(...parseHistory(configHistory))
        } else {
            await writeMissedModal(configPath, 'Config File Not Found', context)
        }
    }

    if (context.whisper.config !== undefined && context.whisper.config !== '') {
        if (!context.whisper.config.toLowerCase().endsWith('.md')) context.whisper.config += '.md'
        const whisperPath = disk.joinPath(context.user.data, 'configs', context.whisper.config)

        if (await disk.pathExists(whisperPath)) {
            let whisperContext: any = {
                file: { path: whisperPath, disk },
            }

            whisperContext = await loadMarkdown(whisperContext as Context)
            configContext.whisper.input = whisperContext.file.text

            whisperContext = deepmerge(configContext, whisperContext)
            context.whisper = deepmerge(context.whisper, whisperContext.whisper)
        } else {
            await writeMissedModal(whisperPath, 'Whisper File Not Found', context)
        }
    }

    return context
}

export async function loadMarkdown(context: Context): Promise<Context> {
    const content = (await context.file.disk.readFile(context.file.path)).trim()

    if (content.startsWith('---\n')) {
        const [frontMatter, ...bodyContent] = content.slice(4).split('\n---')

        context = deepmerge(parseYaml(frontMatter), context)
        context.file.text = bodyContent.join('\n---').trim()
    } else context.file.text = content

    return context
}

export function loadContent(context: Context): Context {
    context.user.history.push(...parseHistory(context))

    for (const message of context.user.history) {
        if (message.role === 'system') {
            message.content = message.content.replace(linkRegex, '!$1')
        }
    }

    if (context.user.history.at(-1)?.role === 'user') {
        context.user.label = context.user.history.at(-1)?.name ?? 'User'
        const content = context.user.history.pop()?.content ?? ''

        context.user.prompt = content.replace(linkRegex, '!$1')
    } else context.user.prompt = ''

    return context
}
