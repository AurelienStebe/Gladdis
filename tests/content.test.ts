import { stringify } from 'yaml'

import { describe, it, expect, afterEach, vi } from 'vitest'

import { diskInterface } from '../src/commands.js'
import { loadContext, loadContent } from '../src/utils/loaders.js'

import type { Context } from '../src/types/context.js'

const complexFile = `
> Quoted Message 1
My User Message 1
---
__User:__ User Message 2
---
__Myself:__ User Message 3
---
__MyFriend:__ User Message 4
---
>__User:__ Quoted Message 2

__System:__ System Message 1
---
__Gladdis:__ Gladdis Message 1
---
__Assistant:__ Gladdis Message 2
---
> """Quoted Message 3"""
"""Text block 1
---
> is escaped"""
---
"""
Text block 2 with
multiline content
and a code block:
\`\`\`bash
echo "Code block with Bash code"
\`\`\`
"""
---
>Quoted Message 4
> with multilines
\`\`\`
alert("Code block with JS code")
\`\`\`
---
\`\`\`python
def test():
    """Test function."""
    print("Hello World!")
\`\`\`
---
__Myself:__![[recording.webm]]

> [!QUOTE]- Transcript from "recording.webm"
> Here is the transcription of the recording.
---
__System:__ [[document.pdf]]

> [!ABSTRACT]- Content from "document.pdf"
> Here is the text content of the PDF file.
---
__Myself:__ <http://mysite.com>

> [!EXAMPLE]- Web Page from <http://mysite.com>
> Here is the main content of the Web Page.
---
__User:__Prompt Text`

describe('the history Content', () => {
    afterEach(() => {
        vi.restoreAllMocks()
    })

    it('parses simple messages', async () => {
        const simpleFile = 'Message 1\n---\nMessage 2\n---\nMessage 3'
        diskInterface.readFile = vi.fn().mockResolvedValue(simpleFile)

        const callContext: any = {
            file: { path: 'path', disk: diskInterface },
            user: { env: { NODE_ENV: 'test' } },
        }

        const context = loadContent(await loadContext(callContext as Context))

        expect(context.user.prompt).toBe('Message 3')
        expect(context.user.history).toStrictEqual([
            {
                role: 'user',
                content: 'Message 1',
            },
            {
                role: 'user',
                content: 'Message 2',
            },
        ])
    })

    it('parses complex messages', async () => {
        diskInterface.readFile = vi.fn().mockResolvedValue(complexFile)

        const callContext: any = {
            file: { path: 'path', disk: diskInterface },
            user: { env: { GLADDIS_DEFAULT_USER: 'Myself' } },
        }

        const context = loadContent(await loadContext(callContext as Context))

        expect(context.user.prompt).toBe('Prompt Text')
        expect(context.user.history).toMatchSnapshot()
    })

    it('defaults to System in config', async () => {
        const inputFile = `---\n${stringify({
            gladdis: { config: 'gladdis' },
            whisper: { config: 'whisper' },
        })}---\n`

        const gladdisConf = 'System Prompt 1\n---\nSystem Prompt 2'
        const whisperConf = 'Whisper input is not parsed\n---\n> !'

        diskInterface.pathExists = vi.fn().mockResolvedValue(true)

        diskInterface.readFile = vi
            .fn()
            .mockResolvedValueOnce(inputFile)
            .mockResolvedValueOnce(gladdisConf)
            .mockResolvedValueOnce(whisperConf)

        const callContext: any = {
            file: { path: 'path', disk: diskInterface },
            user: { env: { GLADDIS_DEFAULT_USER: 'Myself' } },
        }

        const context = loadContent(await loadContext(callContext as Context))

        expect(context.whisper.input).toBe(whisperConf)
        expect(context.user.history).toStrictEqual([
            {
                role: 'system',
                content: 'System Prompt 1',
            },
            {
                role: 'system',
                content: 'System Prompt 2',
            },
        ])
    })
})
