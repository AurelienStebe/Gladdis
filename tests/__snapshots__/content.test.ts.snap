// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`the history Content > parses complex messages 1`] = `
[
  {
    "content": "My User Message 1",
    "name": "Myself",
    "role": "user",
  },
  {
    "content": "User Message 2",
    "role": "user",
  },
  {
    "content": "User Message 3",
    "name": "Myself",
    "role": "user",
  },
  {
    "content": "User Message 4",
    "name": "MyFriend",
    "role": "user",
  },
  {
    "content": "System Message 1",
    "role": "system",
  },
  {
    "content": "Gladdis Message 1",
    "role": "assistant",
  },
  {
    "content": "Gladdis Message 2",
    "role": "assistant",
  },
  {
    "content": """"Text block 1
---
> is escaped"""",
    "name": "Myself",
    "role": "user",
  },
  {
    "content": """"
Text block 2 with
multiline content
and a code block:
\`\`\`bash
echo "Code block with Bash code"
\`\`\`
"""",
    "name": "Myself",
    "role": "user",
  },
  {
    "content": "\`\`\`
alert("Code block with JS code")
\`\`\`",
    "name": "Myself",
    "role": "user",
  },
  {
    "content": "\`\`\`python
def test():
    """Test function."""
    print("Hello World!")
\`\`\`",
    "name": "Myself",
    "role": "user",
  },
  {
    "content": ""Here is the transcription of the recording." (transcribed and read)",
    "name": "Myself",
    "role": "user",
  },
  {
    "content": ""document.pdf":
"""
Here is the text content of the PDF file.
"""

",
    "role": "system",
  },
  {
    "content": "@"https://mysite.com"
"""
Here is the text content of the web page.
"""

",
    "name": "Myself",
    "role": "user",
  },
]
`;
